packages/core/src/modules/mdoc/Mdoc.ts

```diff
@@ -0,0 +1,121 @@
+import type { MdocCreateOptions, MdocNameSpaces, MdocVerifyOptions } from './MdocOptions'
+import type { AgentContext } from '../../agent'
+import type { IssuerSignedDocument } from '@protokoll/mdoc-client'
+
+import { DeviceSignedDocument, Document, Verifier, cborEncode, parseIssuerSigned } from '@protokoll/mdoc-client'
+
+import { getJwkFromKey, JwaSignatureAlgorithm } from '../../crypto'
+import { X509Certificate, X509ModuleConfig } from '../x509'
+
+import { TypedArrayEncoder } from './../../utils'
+import { getMdocContext } from './MdocContext'
+import { MdocError } from './MdocError'
+
+/**
+ * This class represents a IssuerSigned Mdoc Document,
+ * which are the actual credentials being issued to holders.
+ */
+export class Mdoc {
+  public base64Url: string
+  private constructor(private issuerSignedDocument: IssuerSignedDocument) {
+    const issuerSigned = issuerSignedDocument.prepare().get('issuerSigned')
+    this.base64Url = TypedArrayEncoder.toBase64URL(cborEncode(issuerSigned))
+  }
+
+  public static _interalFromIssuerSignedDocument(issuerSignedDocument: IssuerSignedDocument) {
+    return new Mdoc(issuerSignedDocument)
+  }
+
+  public static fromBase64Url(mdocBase64Url: string, expectedDocType?: string): Mdoc {
+    const issuerSignedDocument = parseIssuerSigned(TypedArrayEncoder.fromBase64(mdocBase64Url), expectedDocType)
+    return new Mdoc(issuerSignedDocument)
+  }
+
+  public get docType(): string {
+    return this.issuerSignedDocument.docType
+  }
+
+  public get alg(): JwaSignatureAlgorithm {
+    const algName = this.issuerSignedDocument.issuerSigned.issuerAuth.algName
+    if (!algName) {
+      throw new MdocError('Cannot extract the signature algorithm from the mdoc.')
+    }
+    if (Object.values(JwaSignatureAlgorithm).includes(algName as JwaSignatureAlgorithm)) {
+      return algName as JwaSignatureAlgorithm
+    }
+    throw new MdocError(`Cannot parse mdoc. The signature algorithm '${algName}' is not supported.`)
+  }
+
+  public get validityInfo() {
+    return this.issuerSignedDocument.issuerSigned.issuerAuth.decodedPayload.validityInfo
+  }
+
+  public get deviceSignedNamespaces(): MdocNameSpaces {
+    if (this.issuerSignedDocument instanceof DeviceSignedDocument === false) {
+      throw new MdocError(`Cannot get 'device-namespaces from a IssuerSignedDocument. Must be a DeviceSignedDocument.`)
+    }
+
+    return this.issuerSignedDocument.allDeviceSignedNamespaces
+  }
```

Comment:
isn't device signed the presentation?

Reply:
yes. It can also include device-signed namespaces. Essentially self-attested claims.

Reply:
but shouldn't this be on the mdoc presentation then?

Reply:
A presentation is a set of mdoc's. If we want to go that route we need to split Mdoc class up into IssuerSigned / Device signed document class.